feat: Initial Code Task Runners support (no-changelog)

valya opened the pull request:
## Summary

Implement initial Task Runner support (previously Agents). This PR contains a large chunk of the work, but there's still some more work to do before this can be used by an average user.

## Related Linear tickets, Github issues, and Community forum posts

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
-->

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

Commit:
wip: Add manager and ws server

Commit:
Merge remote-tracking branch 'origin/master' into pay-1850-agent-manager-in-main-instances

Commit:
Merge remote-tracking branch 'origin/master' into pay-1850-agent-manager-in-main-instances

Commit:
wip

Commit:
wip

Commit:
Merge remote-tracking branch 'origin/master' into pay-1850-agent-manager-in-main-instances

Commit:
Basic execution working

Missing RPC for node execute function and console.log, but the
workflow data proxy is working!

Commit:
Merge remote-tracking branch 'origin/master' into pay-1850-agent-manager-in-main-instances

Commit:
RPC working!

MarcL commented on packages/@n8n/config/src/configs/agents.config.ts:
```diff
@@ -0,0 +1,11 @@
+
+@Config
+export class AgentsConfig {
+	// Defaults to false for now
```
```suggestion
	// Defaults to true for now
```
The comment lies! 🤣

ivov committed:
refactor(core): Prepare agents for testing (#10738)

tomi commented:
Couple general comments about the implementation:
- Could we please prefer `Map`s over plain JS objects when they are used for storing key-value pairs? `Map`s are more type safe (since we don't have `noUncheckedIndexedAccess` specified) and the API is more ergonomic for KV use.
- You're using optional chaining in the function call (e.g. `this.requesters[task.requesterId]?.({...})`) Is this intended or are we hiding possible errors? Better approach would be to use `assert`s to validate invariants.

Commit:
move task runner into main repo

Commit:
clean up Node.js task-runner code

ivov committed:
test(core): Tests for task broker (#10749)

Commit:
Merge remote-tracking branch 'origin/master' into pay-1850-agent-manager-in-main-instances

Commit:
Merge remote-tracking branch 'origin/pay-1850-agent-manager-in-main-instances' into pay-1850-agent-manager-in-main-instances

Commit:
add customData processing

fixed a lot of linting issues too

ivov committed:
test(core): More task broker tests (#10871)

Commit:
Merge branch 'pay-1850-agent-manager-in-main-instances' of github.com:n8n-io/n8n into pay-1850-agent-manager-in-main-instances

ivov committed:
feat(core): Support browser and server logging in task runners (#10949)

tomi committed:
feat: Implement auth for task runners (#10937)

Commit:
Merge remote-tracking branch 'origin/master' into pay-1850-agent-manager-in-main-instances

Commit:
add n8n uri env var for the task runner

also fix eslint issues

tomi commented on packages/cli/src/runners/task-broker.service.ts:
```diff
@@ -0,0 +1,556 @@
+
+@Service()
+export class TaskBroker {
+	private knownRunners: Record<
```
Same comment here about the `Map` vs `Record`

tomi commented on packages/cli/src/runners/task-managers/task-manager.ts:
```diff
@@ -0,0 +1,373 @@
+	data: TaskData;
+}
+
+export class JobError extends ApplicationError {}
```
Could be moved to own file

tomi commented on packages/cli/src/runners/task-managers/task-manager.ts:
```diff
@@ -0,0 +1,373 @@
+};
+
+export class TaskManager {
+	requestAcceptRejects: Record<string, { accept: RequestAccept; reject: RequestReject }> = {};
```
Also here: `Map`

tomi commented on packages/cli/src/runners/task-managers/task-manager.ts:
```diff
@@ -0,0 +1,373 @@
+			});
+
+			const resultData = await dataPromise;
+			if (resultData.customData) {
+				Object.entries(resultData.customData).forEach(([k, v]) => {
+					if (!runExecutionData.resultData.metadata) {
+						runExecutionData.resultData.metadata = {};
+					}
+					runExecutionData.resultData.metadata[k] = v;
+				});
+			}
```
What is this `customData` stuff? Could we have a comment explaining it

valya replied:
I've added a comment now, let me know if it's not clear enough.

tomi reviewed (commented):
Good stuff 👌 Many of the comments are clarifying questions or about error scenarios. We can also add follow-up tickets for most of them so we can get this long living branch merged

tomi committed:
fix: Deregister task runner when connection closes (#11006)

Commit:
lint fixes and fix `$execution.customData`

I'm not 100% sure why customData even worked before? It shouldn't have
been able to but I did test it.

Commit:
Merge branch 'pay-1850-agent-manager-in-main-instances' of github.com:n8n-io/n8n into pay-1850-agent-manager-in-main-instances

Commit:
Merge remote-tracking branch 'origin/master' into pay-1850-agent-manager-in-main-instances

valya commented on packages/@n8n/task-runner-node-js/src/code.ts:
```diff
@@ -0,0 +1,130 @@
+		};
+
+		const result = (await runInNewContext(
+			`module.exports = async function() {${task.settings!.code}\n}()`,
```
This is the exact same string we already use, so I just went with it. And no, no infinite loop protection. In theory we might be able to use worker threads since we're processing JSON serialised objects anyway.

valya commented on packages/@n8n/task-runner-node-js/src/task-runner.ts:
```diff
@@ -0,0 +1,364 @@
+	}
+
+	send(message: RunnerMessage.ToN8n.All) {
+		this.ws.send(JSON.stringify(message));
```
Sorry, not sure which part you're pointing to here?

tomi replied:
I mean, the `.send` method [has a callback](https://github.com/websockets/ws/blob/HEAD/doc/ws.md#websocketsenddata-options-callback) that is called with an error if the send fails. Just thinking if we should handle those errors somehow, e.g. by logging them

valya commented on packages/@n8n/task-runner-node-js/src/task-runner.ts:
```diff
@@ -0,0 +1,364 @@
+				this.taskCancelled(message.taskId);
+				break;
+			case 'broker:tasksettings':
+				void this.receivedSettings(message.taskId, message.settings);
```
Anything that could error in `receivedSettings` is wrapped in a try/catch, so it should be fine

valya commented on packages/@n8n/task-runner-node-js/src/task-runner.ts:
```diff
@@ -0,0 +1,364 @@
+
+	processDataResponse(requestId: string, data: unknown) {
+		const request = this.dataRequests[requestId];
+		if (!request) {
```
Basically never, it's basically just there for typechecking

valya replied:
In other places (like the broker and manager), they should basically never happen. It's more just in case something else goes wrong (busy event loop), and it's getting messages that are already resolved elsewhere. I'd considered logging them, but I thought it'd either never get hit, or be really noisy in some environments.

valya commented on packages/@n8n/task-runner-node-js/tsconfig.json:
```diff
@@ -0,0 +1,17 @@
+		"tsBuildInfoFile": "dist/typecheck.tsbuildinfo",
+		// TODO: remove all options below this line
+		"strict": false,
+		"useUnknownInCatchVariables": false
```
This was because I'd just copied the tsconfig files over from `packages/cli`, removed.

valya commented on packages/cli/src/runners/__tests__/task-broker.test.ts:
```diff
@@ -0,0 +1,501 @@
+				runnerId: 'runner1',
+				taskType: 'taskType1',
+				validFor: 1000,
+				validUntil: now + BigInt(1000 * 1_000_000),
```
Where would be best to add a function like this? I know we have `constants.ts`, but do we have a good place to put these kinds of functions?

tomi replied:
It could be in the test file itself for example

Commit:
fix record->map issue in test

tomi commented on packages/@n8n/task-runner-node-js/src/code.ts:
```diff
@@ -44,11 +62,12 @@ export interface AllData {
-const getAdditionalKeys = (): IWorkflowDataProxyAdditionalKeys => {
-	return {};
-};
+// const getAdditionalKeys = (): IWorkflowDataProxyAdditionalKeys => {
+// 	return {};
+// };
```
Commented out code

tomi commented on packages/@n8n/task-runner-node-js/src/task-runner.ts:
```diff
@@ -102,8 +100,8 @@ export class TaskRunner {
 
 	deleteStaleOffers() {
 		for (const key of Object.keys(this.openOffers)) {
```
nit: I think this can be done without using `Object.keys`
```suggestion
		this.openOffers.forEach((offer, key) => {
		  if (offer.validUntil < process.hrtime.bigint()) {
				this.openOffers.delete(key);
			}
		})
```

tomi commented on packages/cli/src/runners/__tests__/task-broker.test.ts:
```diff
@@ -60,13 +61,13 @@ describe('TaskBroker', () => {
 
-			expect(knownRunners[runnerId].runner).toEqual(runner);
-			expect(knownRunners[runnerId].messageCallback).toEqual(messageCallback);
+			expect(knownRunners.get(runnerId)!.runner).toEqual(runner);
+			expect(knownRunners.get(runnerId)!.messageCallback).toEqual(messageCallback);
```
nit: Could use `?` here instead of `!`. In that case we get `AssertionError` from jest with proper formatting, where as with `!` we get `ReferenceError` from runtime if the `knownRunners` don't have `runnerId`

tomi commented on packages/cli/src/runners/task-broker.service.ts:
```diff
@@ -90,35 +87,35 @@ export class TaskBroker {
-			delete this.knownRunners[runnerId];
+		if (this.knownRunners.has(runnerId)) {
+			this.knownRunners.delete(runnerId);
 		}
```
nit: can also delete directly, as it's no-op if the key is not in the map

```suggestion
	this.knownRunners.delete(runnerId);
```

tomi commented on packages/cli/src/runners/task-broker.service.ts:
```diff
@@ -90,35 +87,35 @@ export class TaskBroker {
-			delete this.requesters[requesterId];
+		if (this.requesters.has(requesterId)) {
+			this.requesters.delete(requesterId);
 		}
```
nit: same here
```suggestion
	this.requesters.delete(requesterId);
```

tomi commented on packages/cli/src/runners/task-managers/task-manager.ts:
```diff
@@ -199,87 +219,90 @@ export class TaskManager {
 		switch (message.type) {
 			case 'broker:taskready':
-				this.jobReady(message.requestId, message.taskId);
+				this.taskReady(message.requestId, message.taskId);
```
👍 for renaming these

tomi reviewed (dismissed):
Looks good! Couple nitpick comments but IMO ready to be merged 👏

tomi committed:
Review fixes

janober commented:
Got released with [`n8n@1.62.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.62.1)
